### PR TITLE
[JENKINS-29796] Multiple refspecs should be matched using OR, not AND

### PR DESCRIPTION
**Pull request by hannes-ucsc**
https://issues.jenkins-ci.org/browse/JENKINS-29796

### CODE REVIEW

**Comment by MarkEWaite** [comment]
Could you provide a test which shows the problem?  A bug fix without a test is too easy to lose when someone else submits a different change.

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 1]
```diff
@@ -591,13 +591,17 @@ private PollingResult compareRemoteRevisionWithImpl(Job<?, ?> project, Launcher
...
                         for (RefSpec spec : refSpecs) {
-                            if (!spec.matchSource(head)) {
-                                listener.getLogger().println("Ignoring " + head + " as it doesn't match configured refspecs");
-                                it.remove();
+                            if (spec.matchSource(head)) {
+                                match = true;
                                 break;
                             }
                         }
+                        if(! match) {
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/GitSCM.java]: `if (!match) {`

**Comment by hannes-ucsc** [comment]
@MarkEWaite Sure, I'll write a test. Do you agree with the general idea, though? I was a little perplexed as to why the original add AND semantics. It seemed like a deliberate decision and I'm worried I'm wasting my time because I don't see the big picture.

**Comment by MarkEWaite** [comment]
I think you're on the right track.  I don't know why multiple refspecs would all be required to match.  That seems counter to my expectation of the cases where I might use multiple refspecs.

**Comment by hannes-ucsc** [comment]
Added a test. The test fails on the master branch and succeeds on this one.

The tests in SCMTriggerTest are a bit wonky IMHO. For example, the `commits` parameter to `SCMTriggerTest.check()` is never used. I was also surprised to see the `poll.hasChanges()` in `check()` is always false.

**Comment by MarkEWaite** [comment]
It is a complicated enough plugin that I often discover unexpected twists and turns while writing tests. That discovery process has been part of the fun of trying to maintain the plugin.  Thanks for writing the test.

**Pull request merged**